### PR TITLE
Remove Git backend + Improve API + Cache functor

### PR DESCRIPTION
**Pull request by smondet**
This PR goes on top of  https://github.com/smondet/trakeva/pull/12; it contains 3 distinct steps:

- branch https://github.com/smondet/trakeva/tree/remove-git-backend removes the `git_commands` backend.
- branch https://github.com/smondet/trakeva/tree/get-all-keys-semantics changes the semantics of `get_all` and `iterator` (easier without the git brackend)
- branch https://github.com/smondet/trakeva/tree/cache_functor add a greedy in-memory cache functor (easier to the new semantics)

[Review on Reviewable](https://reviewable.io/reviews/smondet/trakeva/13)

### CODE REVIEW

File: src/lib/trakeva_cache.ml [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+let debug = ref false
+
+let dbg fmt = ksprintf (eprintf "Trakeva_cache: %s\n%!") fmt
+    
+module Greedy_cache = struct
+
+  type t = {
+    collections: (string, (string, string) Hashtbl.t) Hashtbl.t;
+  }
```
Comment (rleonid) [inline src/lib/trakeva_cache.ml]: What's the purpose of the extra boxing? (ie. why wrap it in a record?)

File: src/lib/trakeva_cache.ml [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    Hashtbl.add col key value
+
+  let fold_collection col ~init ~f =
+    let r = ref init in
+    Hashtbl.iter (fun key value ->
+        if !debug then dbg "fold_collection %s, %s" key value;
+        r := f !r ~key ~value) col;
+    !r
+
+
```
Comment (rleonid) [inline src/lib/trakeva_cache.ml]: These names (`from_collection`, `new_collection`, `add_to_collection` and `fold_collection`) seem redundant. Why not rename the module and just have `find`, `new`, `add`, and `fold` ?

File: src/lib/trakeva_cache.ml [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+open Nonstd
+open Pvem_lwt_unix
+open Pvem_lwt_unix.Deferred_result
+module String = StringLabels
+
+let debug = ref false
+
+let dbg fmt = ksprintf (eprintf "Trakeva_cache: %s\n%!") fmt
+    
+module Greedy_cache = struct
```
Comment (rleonid) [inline src/lib/trakeva_cache.ml]: `Greedy_cache` is a bit confusing here. The `Hashtbl` grows without bounds, so I'm not certain what's `greedy` about it.

File: src/lib/trakeva_cache.ml [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    cache: Greedy_cache.t;
+    mutex: Lwt_mutex.t;
+  }
+  let load s =
+    KV_DB.load s
+    >>= fun original ->
+    return {original; cache = Greedy_cache.create (); mutex = Lwt_mutex.create ()}
+
+  let close {original; cache} =
+    KV_DB.close original
```
Comment (rleonid) [inline src/lib/trakeva_cache.ml]: The `cache` in `{original; cache}` is superfluous. Were you going to do something with it?

File: src/lib/trakeva_cache.ml [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    let all = ref None in
+    begin fun () ->
+      begin
+        begin match !all with
+        | None ->
+          get_all t ~collection
+          >>= fun all_keys ->
+          all := Some all_keys;
+          return all_keys
+        | Some l -> return l
```
Comment (rleonid) [inline src/lib/trakeva_cache.ml]: This looks awkward. Can't you move this block outside of the `begin fun () ->` ? Or is this meant to be lazy.

**Comment by rleonid** [comment]
Finished first pass. Looks fine.

File: src/lib/trakeva_cache.ml (hunk 1 above)
Comment (smondet) [inline src/lib/trakeva_cache.ml]: To force the typing + to add things later.

File: src/lib/trakeva_cache.ml (hunk 2 above)
Comment (smondet) [inline src/lib/trakeva_cache.ml]: Yes but the `t` is still the greedy cache in those functions.

I will think about reshaping that API though (e.g. `fold_collection` is missing the `t`).

File: src/lib/trakeva_cache.ml (hunk 3 above)
Comment (smondet) [inline src/lib/trakeva_cache.ml]: Yes it grows without bounds and it eats a lot at once.
I use [greedy](http://dictionary.reference.com/browse/greedy?s=t) a synonym of “gluttonous.”

File: src/lib/trakeva_cache.ml (hunk 4 above)
Comment (smondet) [inline src/lib/trakeva_cache.ml]: OK fixed in f42c0a4

File: src/lib/trakeva_cache.ml (hunk 5 above)
Comment (smondet) [inline src/lib/trakeva_cache.ml]: Yes it is a “staged” function (it creates a stream) so we need the `let all = ref None` before the `fun () ->`.

**Comment by smondet** [comment]
Merged into master

**Pull request closed without merging**